gadget/device_linux.go

```diff
@@ -45,7 +45,15 @@ func FindDeviceForStructure(ps *LaidOutStructure) (string, error) {
 	var candidates []string
 
 	if ps.Name != "" {
-		byPartlabel := filepath.Join(dirs.GlobalRootDir, "/dev/disk/by-partlabel/", encodeLabel(ps.Name))
+		// if the device is MBR, we have to use by-label, as MBR does not have
+		// by-partlabel
+		var byPartlabel string
+		if ps.EffectiveSchema() == GPT {
+			byPartlabel = filepath.Join(dirs.GlobalRootDir, "/dev/disk/by-partlabel/", encodeLabel(ps.Name))
+		} else {
+			byPartlabel = filepath.Join(dirs.GlobalRootDir, "/dev/disk/by-label/", encodeLabel(ps.Name))
```

Comment:
I think this is misplaced here. Looked at `ubuntu-image` again. The logic there uses partition name as a fallback when no filesystem label is provided regardless of GPT/MBR. Perhaps we should have the following:
```
if ps.Name != "" {
	byPartlabel := filepath.Join(dirs.GlobalRootDir, "/dev/disk/by-partlabel/", encodeLabel(ps.Name))
	candidates = append(candidates, byPartlabel)
}
if ps.HasFilesystem() {
	fsLabel := ps.EffectiveFilesystemLabel()
	if fsLabel == "" && ps.Name != "" {
		fsLabel = ps.Name
	}
	byFsLabel := filepath.Join(dirs.GlobalRootDir, "/dev/disk/by-label/", encodeLabel(fsLabel))
	candidates = append(candidates, byFsLabel)
}
```
and there's no longer a need to have `VolumeStructure.EffectiveSchema()` anymore

Reply:
yes this sounds like the proper way to do this

Reply:
Fixed